### PR TITLE
Add PoseProgressChecker

### PR DESCRIPTION
**Pull request by doisyg**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Dexory robot |

---

## Description of contribution in a few bullet points

This PR adds a `nav2_controller::PoseProgressChecker` inheriting from the `nav2_controller::SimpleProgressChecker` which is itself a `nav2_core::ProgressChecker`. 
Similarly as `nav2_core::GoalChecker`/ `nav2_controller::SimpleGoalChecker` / `nav2_controller::StoppedGoalChecker` 

This new `PoseProgressChecker` builds on the behavior of the `SimpleProgressChecker` by adding a new parameter `required_movement_angle`. Allowing the plugin to considers there is still progress when there is no translation movement, from the moment there is a rotation movement superior to `required_movement_angle` within the `movement_time_allowance`.

Note: `pose_distance` from `SimpleProgressChecker` was changed to `SimpleProgressChecker::pose_distance` and declared in the header file in order to be used by the inherited class.

Appropriate tests were added (along with a reset test for the StoppedGoalChecker which seemed missing).

## Description of documentation updates required from your changes
 
Add entry here : https://github.com/ros-planning/navigation.ros.org/tree/master/configuration/packages/nav2_controller-plugins

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_controller/plugins/test/progress_checker.cpp [hunk 1]
```diff
@@ -119,12 +124,49 @@ TEST(SimpleProgressChecker, unit_tests)
...
+  checkMacro(rpc, 0, 0, 0, 0, 0, -1, 11, true);
+  // 11s translation
+  checkMacro(rpc, 0, 0, 0, 1, 0, 0, 11, true);
+  checkMacro(rpc, 0, 0, 0, 0, 1, 0, 11, true);
+  // 11s rotation
+  checkMacro(rpc, 0, 0, 0, 0, 0, 1, 11, true);
+  checkMacro(rpc, 0, 0, 0, 0, 0, -1, 11, true);
+  // 11s no movement
+  checkMacro(rpc, 0, 0, 0, 0, 0, 0, 11, false);
+
```
Comment (SteveMacenski) [inline nav2_controller/plugins/test/progress_checker.cpp:168]: Linty lint linter

File: nav2_controller/include/nav2_controller/plugins/rotation_progress_checker.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    const std::string & plugin_name) override;
+  bool check(geometry_msgs::msg::PoseStamped & current_pose) override;
+
+protected:
+  /**
+   * @brief Calculates robots movement from baseline pose
+   * @param pose Current pose of the robot
+   * @return true, if movement is greater than radius_, or false
+   */
+  bool is_robot_moved_enough(const geometry_msgs::msg::Pose2D & pose);
```
Comment (SteveMacenski) [inline nav2_controller/include/nav2_controller/plugins/rotation_progress_checker.hpp:46]: Use `camelCase` for methods

File: nav2_controller/include/nav2_controller/plugins/rotation_progress_checker.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  static double pose_angle_distance(
```
Comment (SteveMacenski) [inline nav2_controller/include/nav2_controller/plugins/rotation_progress_checker.hpp:48]: ditto

File: nav2_controller/plugins/rotation_progress_checker.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+#include "nav2_util/node_utils.hpp"
+#include "pluginlib/class_list_macros.hpp"
+
+using rcl_interfaces::msg::ParameterType;
+using std::placeholders::_1;
+
+namespace nav2_controller
+{
+
+void RotationProgressChecker::initialize(
```
Comment (SteveMacenski) [inline nav2_controller/plugins/rotation_progress_checker.cpp:33]: Maybe rename to `PoseProgressChecker` to imply it is both doing translational and rotational? `Rotation` implies it only cares about rotation when its actually doing both. I'm also open to other naming suggestions

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally LGTM just peanuts and then the usual documentation updates

File: nav2_controller/plugins/rotation_progress_checker.cpp (hunk 4 above)
Comment (doisyg) [reply nav2_controller/plugins/rotation_progress_checker.cpp]: `PoseProgressChecker` sounds nicely

**Comment by doisyg** [comment]
Documentation: https://github.com/ros-planning/navigation.ros.org/pull/414

Note sure why the CI is failing. Maybe because `PoseProgressChecker.unit_tests ` is too long (80s locally) ? Not the most efficient approach but I just mimicked `SimpleProgressChecker.unit_tests `.

File: nav2_controller/plugins/pose_progress_checker.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+bool PoseProgressChecker::check(geometry_msgs::msg::PoseStamped & current_pose)
+{
+  // relies on short circuit evaluation to not call is_robot_moved_enough if
+  // baseline_pose is not set.
+  geometry_msgs::msg::Pose2D current_pose2d;
+  current_pose2d = nav_2d_utils::poseToPose2D(current_pose.pose);
+
+  if ((!baseline_pose_set_) || (PoseProgressChecker::isRobotMovedEnough(current_pose2d))) {
```
Comment (SteveMacenski) [inline nav2_controller/plugins/pose_progress_checker.cpp:57]:
```suggestion
  if (!baseline_pose_set_ || PoseProgressChecker::isRobotMovedEnough(current_pose2d)) {
```

File: nav2_controller/plugins/pose_progress_checker.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    resetBaselinePose(current_pose2d);
+    return true;
+  }
+  return !((clock_->now() - baseline_time_) > time_allowance_);
```
Comment (SteveMacenski) [inline nav2_controller/plugins/pose_progress_checker.cpp:61]:
```suggestion
  return clock_->now() - baseline_time_ <= time_allowance_;
```

Seems simpler

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
@doisyg I looked at the logs in `nav2_controller/stdout.log` in the artifacts:

```
6: Test command: /usr/bin/python3.10 "-u" "/opt/ros/rolling/share/ament_cmake_test/cmake/run_test.py" "/opt/overlay_ws/test_results/nav2_controller/pctest.gtest.xml" "--package-name" "nav2_controller" "--output-file" "/opt/overlay_ws/build/nav2_controller/ament_cmake_gtest/pctest.txt" "--command" "/opt/overlay_ws/build/nav2_controller/plugins/test/pctest" "--gtest_output=xml:/opt/overlay_ws/test_results/nav2_controller/pctest.gtest.xml"
6: Test timeout computed to be: 60
6: -- run_test.py: invoking following command in '/opt/overlay_ws/build/nav2_controller/plugins/test':
6:  - /opt/overlay_ws/build/nav2_controller/plugins/test/pctest --gtest_output=xml:/opt/overlay_ws/test_results/nav2_controller/pctest.gtest.xml
6: [==========] Running 4 tests from 2 test suites.
6: [----------] Global test environment set-up.
6: [----------] 2 tests from SimpleProgressChecker
6: [ RUN      ] SimpleProgressChecker.progress_checker_reset
6: [INFO] [1681259294.560951641] [progress_checker]: 
6: 	progress_checker lifecycle node launched. 
6: 	Waiting on external lifecycle transitions to activate
6: 	See https://design.ros2.org/articles/node_lifecycle.html for more information.
6: [INFO] [1681259294.561037439] [progress_checker]: Destroying
6: [       OK ] SimpleProgressChecker.progress_checker_reset (9 ms)
6: [ RUN      ] SimpleProgressChecker.unit_tests
6: [INFO] [1681259294.566507040] [progress_checker]: 
6: 	progress_checker lifecycle node launched. 
6: 	Waiting on external lifecycle transitions to activate
6: 	See https://design.ros2.org/articles/node_lifecycle.html for more information.
6: [INFO] [1681259330.567127812] [progress_checker]: Destroying
6: [       OK ] SimpleProgressChecker.unit_tests (36007 ms)
6: [----------] 2 tests from SimpleProgressChecker (36016 ms total)
6: 
6: [----------] 2 tests from PoseProgressChecker
6: [ RUN      ] PoseProgressChecker.pose_progress_checker_reset
6: [INFO] [1681259330.573627533] [pose_progress_checker]: 
6: 	pose_progress_checker lifecycle node launched. 
6: 	Waiting on external lifecycle transitions to activate
6: 	See https://design.ros2.org/articles/node_lifecycle.html for more information.
6: [INFO] [1681259330.573667573] [pose_progress_checker]: Destroying
6: [       OK ] PoseProgressChecker.pose_progress_checker_reset (8 ms)
6: [ RUN      ] PoseProgressChecker.unit_tests
6: [INFO] [1681259330.581185393] [pose_progress_checker]: 
6: 	pose_progress_checker lifecycle node launched. 
6: 	Waiting on external lifecycle transitions to activate
6: 	See https://design.ros2.org/articles/node_lifecycle.html for more information.
6/8 Test #6: pctest ...........................***Timeout  60.01 sec
```

That would be the case

**Comment by doisyg** [comment]
I rewrote more complete and faster tests, let's see if it passes the CI

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Thanks!

**Pull request merged**